### ISSUE TITLE
Add password complexity restrictions

### ISSUE THREAD

**Issue by maxxer** [2amigos/yii2-usuario#339]
An admin should be able to impose restrictions on password complexity. For example:

1. require one (or more) capital
1. require one (or more) number
1. require one (or more) special character
1. forbid common  words in password (username, email user portion)
1. ... others?

The easiest approach would be to add a configuration parameter with a custom regexp. So an admin can set (almost) all the rules above with a single option.

Another option is to set a config parameter for each of the above cases (considering what we think it's worth to implement). 

Discussion is open, comments welcome.

**Comment by kartik-v** [comment]
Had developed a [password strength validator extension](https://github.com/kartik-v/yii2-password) for this which can be used. It also has a password strength meter - [docs & demo](https://demos.krajee.com/password).

**Comment by tonydspaniard** [comment]
@maxxer I am not sure about this option. To be honest, I find it terribly annoying for a user to follow the password rules of an app. Whilst I think is a good thing to inform the user that his password is weak (which anybody could do by adding their preferred plugin on the registration view and override it)  I really do not know to force them to create X factor password nor injecting more external dependencies to the module. 

@kartik-v thanks for the reference.

**Comment by BillHeaton** [comment]
I agree. The rules quoted date back to Microsoft in the early 90s which I helped develope and reflect the limitations of processor speeds of that era.  We've come a long way since then in calculating password strength.   

The Microsoft rules are brute forceable in a relatively small amount of time.

Better to encourage longer passwords that are easier to remember.  My method is look at a well known object.  Describe it in four words.  Use that description for the password.  (i.e. "GreyTissueFauxFloral" which is a description of the box of tissue on my desk) 

The current magic number is about 16 characters.  Using rules above are gravy. Intentional misspelling of words is gravy (i.e. "Faux" for "Box").

> @maxxer I am not sure about this option. To be honest, I find it terribly annoying for a user to follow the password rules of an app. Whilst I think is a good thing to inform the user that his password is weak (which anybody could do by adding their preferred plugin on the registration view and override it) I really do not know about forcing them to create X factor password nor injecting more external dependencies to the module.
> 
> @kartik-v thanks for the reference.

**Comment by mgrabow1** [comment]
For me the strength validator from @kartik-v did the job almost perfectly.

There's only one problem - automatically generated passwords don't meet my strength criteria :)

It would be very useful if one could set (e.g. through some configuration parameters) minimum length and character sets for the generated passwords. Or at least the possibility to extend and override `SecurityHelper` class - that would be enough, because `generatePassword` function is pretty simple.

**Comment by edegaudenzi** [comment]
After 6 years and a few months, I stumbled on this very issue. Fun fact: I've been using the [kartik-v/yii2-password](https://github.com/kartik-v/yii2-password) alongside the yii2-usuario module. Only later I came here  to discover @kartik-v in person was literally the first answer! _(I must say man, you are a legend!)_

Anyhow, this issue literally represents my current real-life situation and it is for this reason I want to share what I've managed to implement; maybe it'll be useful to someone else. Long-story-short: I've done all the necessary to achieve most of the points listed by @maxxer, exploiting the already existing [minPasswordRequirements](https://yii2-usuario.readthedocs.io/en/latest/install/configuration-options/#minpasswordrequirements-type-array-default-lower-gt-1-digit-gt-1-upper-gt-1) which as per version `1.6.3` is only used when generating a password.

Overall architecture of my solution is to create a Yii2-compliant validator and use it to validate the various yii2-usuario forms like `RecoveryForm` and `SettingsForm`. If you are interested, I can pop here a PR.

This is how it looks like in my local environment:
<img width="560" height="308" alt="Image" src="https://github.com/user-attachments/assets/c0d10b17-e7fe-4c9d-bdc8-a2c2cf509992" />
